left_sidebar: Add option to hide summary unread counters.

akarsh-jain-790 opened the pull request:
Adds a user-configurable option to hide unread counters in the left sidebar. Users now have the flexibility to choose whether they want these counters displayed or hidden, reducing distraction for those who find them disruptive.

Fixes: #28759

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

Show/Hide Unread Counter Option:

Added a "Show unread counter" / "Hide unread counter" option as the second item in the home view menu.

<img width="904" alt="Screenshot 2024-02-22 at 12 03 43 PM" src="https://chat.zulip.org/user_uploads/2/1/IGmdCyZAW3d15-u0sqoEG_qP/Screenshot-2024-03-08-at-11.44.29PM.png">


Preferences Setting:

Added a "Show unread message totals on home view and streams" checkbox under SETTINGS / PREFERENCES > Advanced, just below the "Show unread counts for" setting.

<img width="904" alt="Screenshot 2024-02-22 at 12 03 43 PM" src="https://chat.zulip.org/user_uploads/2/2d/hH-js6pJe_L7kNbRoOH9YcNv/Screenshot-2024-03-08-at-11.37.19PM.png">

When "Show Unread Counter" is selected for "No streams":

<details>
Unread counters will not be visible in the left sidebar for streams.

<img width="904" alt="Screenshot 2024-02-22 at 12 03 43 PM" src="https://chat.zulip.org/user_uploads/2/8f/GTBob2q_cHB-JY6ZValvnuh5/Screenshot-2024-03-08-at-11.55.14PM.png">

When "Hide Unread Counter" is selected:

The counter will be displayed as a dot on both the home view and the STREAMS line.
The actual number will be revealed on hover.

<img width="904" alt="Screenshot 2024-02-22 at 12 03 43 PM" src="https://chat.zulip.org/user_uploads/2/9b/_NtEudpoL69vF7kTStpKfs7J/Screenshot-2024-03-08-at-11.55.34PM.png">


Adjusted the logic to ensure consistency in displaying summary counters in the default left sidebar view and in "more topics"/"more conversations" views.
When "Show unread message totals on home view and streams" is enable:
More Conversations View:

<img width="904" alt="Screenshot 2024-02-22 at 12 03 43 PM" src="https://chat.zulip.org/user_uploads/2/91/JlRMYyymVZ5KihrwUmqC0_uT/Screenshot-2024-03-08-at-11.46.21PM.png">


More Topics View:

<img width="904" alt="Screenshot 2024-02-22 at 12 03 43 PM" src="https://chat.zulip.org/user_uploads/2/1f/xg9v-8FAq7l09AYnUuiOk062/Screenshot-2024-03-08-at-11.45.54PM.png">


When "Show unread message totals on home view and streams" is not enable:
More Conversations View:

<img width="904" alt="Screenshot 2024-02-22 at 12 03 43 PM" src="https://chat.zulip.org/user_uploads/2/8d/ryVqq46o5_flv9ubZWpS6szS/Screenshot-2024-03-08-at-11.46.36PM.png">

More Topics View:

<img width="904" alt="Screenshot 2024-02-22 at 12 03 43 PM" src="https://chat.zulip.org/user_uploads/2/94/ZYq61MHAS5mxpBcXVwScKTyU/Screenshot-2024-03-08-at-11.46.52PM.png">
</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [X] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [X] Explains differences from previous plans (e.g., issue description).
- [X] Highlights technical choices and bugs encountered.
- [X] Calls out remaining decisions and concerns.
- [X] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [X] Each commit is a coherent idea.
- [X] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [X] Visual appearance of the changes.
- [X] Responsiveness and internationalization.
- [X] Strings and tooltips.
- [X] End-to-end functionality of buttons, interactions and flows.
- [X] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
@akarsh-jain-790 Has this PR been updated with the latest #design suggestions? If so, please update the screenshots in the PR description, and post a summary comment here about the PR's status.

akarsh-jain-790 commented:
@alya Yeah, I have update the PR description and is now ready for the initial review.

alya commented:
Things go wrong when views are collapsed:

![Screenshot 2024-03-14 at 12 08 14 AM](https://github.com/zulip/zulip/assets/2090066/c1980928-80ac-4a43-84d7-a4b42e2beb7d)
![Screenshot 2024-03-14 at 12 08 03 AM](https://github.com/zulip/zulip/assets/2090066/8515ca8c-dec0-4605-9a6a-c14949426442)

Please try to test your changes more thoroughly; https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#manual-testing has some tips.

akarsh-jain-790 commented:
@alya I have updated the PR and description, can you please re-review.

timabbott commented on api_docs/changelog.md:
```diff
@@ -19,6 +19,13 @@ clients should check the `zulip_feature_level` field, present in the
 format used by the Zulip server that they are interacting with.
 
 ## Changes in Zulip 9.0
+**Feature level 243**
```
You need a blank line before this -- the page doesn't render properly if you don't. Always test your changes -- it's not have to view these docs (/api/changelog).

timabbott commented on api_docs/changelog.md:
```diff
@@ -19,6 +19,13 @@ clients should check the `zulip_feature_level` field, present in the
+  [`PATCH /realm/user_settings_defaults`](/api/update-realm-user-settings-defaults):
+  Added new `web_left_sidebar_unreads_count_summary` display setting, which allows
+  users to selectively show or hide unread message counters in the left sidebar,
+  providing a customizable and distraction-free interface.
```
This does too much selling of the value of the feature; we generally aim to be very concise. Look at other entries for realm settings for examples.

timabbott commented on web/src/left_sidebar_navigation_area_popovers.js:
```diff
@@ -41,6 +41,19 @@ function register_mark_all_read_handler(event) {
+    const {instance} = event.data;
+    const data = {};
+    const home_view_unread_msg_count = user_settings.web_left_sidebar_unreads_count_summary;
+    data.web_left_sidebar_unreads_count_summary = JSON.stringify(!home_view_unread_msg_count);
```
This is a weird way to construct `data`; and worse than the similar blocks just above; can you follow those better conventions?

akarsh-jain-790 replied:
Thank you for your review. I appreciate your feedback. Regarding the construction of `data` I took reference from a similar block below:

```
$popper.one("click", "#toggle_display_starred_msg_count", () => {
                const data = {};
                const starred_msg_counts = user_settings.starred_message_counts;
                data.starred_message_counts = JSON.stringify(!starred_msg_counts);

                channel.patch({
                    url: "/json/settings",
                    data,
                });
                instance.hide();
            });
```

timabbott replied:
Hmm, maybe the right plan is to refactor those as well; the ones above this seem more sensible.

timabbott commented on web/src/left_sidebar_navigation_area_popovers.js:
```diff
@@ -126,6 +139,13 @@ export function initialize() {
+                "click",
+                "#toggle_display_home_view_msg_count",
+                {instance},
+                register_toggle_display_left_sidebar_messsage_count_handler,
```
Always spell "message" as "message"; `msg` and `messsage` are both not desired.

timabbott commented on web/src/sidebar_ui.ts:
```diff
@@ -53,6 +53,25 @@ export function update_invite_user_option(): void {
+    const $direct_message_header = $("#private_messages_section_header");
+
+    if (!user_settings.web_left_sidebar_unreads_count_summary) {
+        $home_view.addClass("hide_unread_counts");
```
Maybe call this class `hide-unread-count-summary` -- that aligns a lot more closely with the setting name. And we're preferring `-`s in new CSS class names and IDs there days.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -10653,6 +10653,11 @@ paths:
                   example: 2
+                web_left_sidebar_unreads_count_summary:
+                  description: |
+                    Enables users to customize the visibility of unread message counters in the left sidebar.
```
This is missing the "Changes" entry.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -15342,6 +15355,12 @@ paths:
+                          web_left_sidebar_unreads_count_summary:
+                            type: boolean
+                            description: |
+                              Enables users to customize the visibility of unread message counters in the left sidebar.
```
This wording is weird; "Enables" is not what the setting does. Something like:

> Whether the web/desktop application's left sidebar displays grand total unread counts.

timabbott commented on web/styles/left_sidebar.css:
```diff
@@ -796,6 +839,42 @@ li.top_left_scheduled_messages {
+            }
+        }
+    }
 }
```
This CSS refactoring is quite hard to understand; can you use some preparatory commits with proper commit messages to explain it?

akarsh-jain-790 commented:
@timabbott Thanks for the detailed feedback! I've addressed your comments and made the necessary updates to the PR. Let me know if you have any further suggestions.

alya commented:
@timabbott Marking for integration review, since you seem to be reviewing this one? Please feel free to re-delegate.

timabbott commented:
@akarsh-jain-790 any chance you're up for updating this following various refactoring that's been merged?

akarsh-jain-790 commented:
> @akarsh-jain-790 any chance you're up for updating this following various refactoring that's been merged?

@timabbott  I've integrated the merged refactoring updates and the code is ready for your review.

akarsh-jain-790 commented:
@timabbott  I have fixed the test and PR is ready for review!

alya commented:
@akarsh-jain-790 are you up for rebasing this again?

karlstolley commented:
@alya that's possible...it'll be easier to see more clearly once the existing conflicts here have been resolved.

akarsh-jain-790 commented:
> @akarsh-jain-790 are you up for rebasing this again?

sure
